docs/docs/genai/tracing/quickstart/typescript-openai.mdx

```diff
@@ -0,0 +1,369 @@
+            messages.push(aiMsg);
+
+            // If the model requests tool call(s), invoke the function
+            if (aiMsg.tool_calls && aiMsg.tool_calls.length > 0) {
+                for (const toolCall of aiMsg.tool_calls) {
+                    const functionName = toolCall.function.name;
+
+                    if (functionName === "get_weather") {
+                        // Invoke the tool function with the provided arguments
+                        const args = JSON.parse(toolCall.function.arguments);
+                        const toolResult = await getWeather(args.latitude, args.longitude);
+                        messages.push({
+                            "role": "tool",
+                            "tool_call_id": toolCall.id,
+                            "content": String(toolResult),
+                        });
+                    } else {
+                        throw new Error(`Invalid tool returned: ${functionName}`);
+                    }
+                }
+
+                // Second LLM call with tool results
+                const finalResponse = await client.chat.completions.create({
+                    model: "o4-mini",
+                    messages: messages
+                });
+                return finalResponse.choices[0].message.content;
+            }
+            return aiMsg.content;
+        },
+        {
+            name: "run_tool_agent",
+            spanType: mlflow.SpanType.AGENT,
+        }
+    );
+}
+
+```
+
+Now we can run the application.
+
+```typescript
+# Run the tool calling agent
+const question = "What's the weather like in Seattle?";
+await runToolAgent(question);
+```
+
+## Step 4: Explore Traces in the UI
+
+After running the application, you can explore the traces in the MLflow UI.
+
+<ImageBox src="/images/llms/tracing/quickstart/openai-tool-calling-trace-detail.png" alt="Tool Calling Trace" width="90%" />
+
+The trace shows all LLM invocations and tool calls, organized in a tree structure. You can also inspect the timeline breakdown by clicking the timeline icon next to the tree view. This helps you understand where the time is spent in the application.
+
+<ImageBox src="/images/llms/tracing/quickstart/openai-tool-calling-trace-timeline.png" alt="Tool Calling Trace Timeline" width="90%" />
+
+## Step 5: Attach Feedbacks on Traces
+
+As a last step of this quickstart, let's attach a feedback on the generated traces. In the real world development, human feedback is critical to improve the quality of the application.
+
+To add a feedback on a trace, you can open the trace detail page and click the "Add new Assessment" button on the top right. It will open an input form where you can provide a various feedback values and metadata. For example, we can add a feedback called " Quality" with a integer value (1~5), indicating how good the answer is. Then we can also put the detailed rationale behind the score for future reference.
```

Comment:
```suggestion
To add a feedback to a trace, you can open the trace detail page and click the "Add new Assessment" button on the top right. It will open an input form where you can provide various feedback values and metadata. For example, we can add a feedback entry called "Quality" with a integer value (1~5), indicating how good the answer is. We can also put the detailed rationale behind the score for future reference.
```